feat: enable ledger hardware wallet implementation.

dawnseeker8 opened the pull request:
## **Description**

The PR enable ledger hardware wallet support in Metamask mobiles. 

**Ledger pre-requisite:**
- Patch `keyring controller` to allow exposing `EthKeyringController` to support ledger keyring operation
- Adding ledger keyring utils `app/core/Ledger/Ledger.ts` to support ledger operation, such as `add ledger, remove ledger, unlock ledger account, get ledger, connect ledger, open ledger app and close ledger app`
- Add `Permission-BluetoothPeripheral` in IOS pod
- Add self develop ledger keyring library  `@consensys/ledgerhq-metamask-keyring`
- Add ledger offi package `@ledgerhq/react-native-hw-transport-ble` and `react-native-ble-plx`

**Enable ledger features:**
- Pair with Ledger devices via bluetooth, and import ledger accounts into metamask mobile.
- Swap token with ledger account.
- Sending web3 transaction with ledger account. (contract deployment, transfer, authorise etc.)
- Connect Dapp with ledger account and execute rpc operation.
- Sign message with ledger account, supported message type:
  - `eth_sign`
  - `personal_sign`
  - `eth_signTypedDataV4`
- Dispatch sign message event on different stages [`REQUEST_SEND`, `COMPLETE`, `ERROR`] in RPCMethodMiddleware to support ledger sign modal interaction
- Implement required ledger metrics to ledger related screens
- Adding `account_type` metrics for swap operation and account connect component
- Implementing `ExtendedKeyringTypes` to consolidate the keyring type from keyring controller and ledger 
- Replace `fontSize` and `fontWeight`  with `TextVariant` on account label tag related components to adopt theme design




## **Related issues**

Fixes: https://github.com/MetaMask/metamask-mobile/issues/1237

## **Manual testing steps**

Need a full regression tests to test all ledger relative feature using https://metamask.github.io/test-dapp/

## **Screenshots/Recordings**

For detailed of screenshot, please refer to our internal Figma design. we will only list out some screenshot below:

Scan devices
![Screenshot 2023-11-08 at 12 18 48](https://github.com/MetaMask/metamask-mobile/assets/7315988/2eb3e3eb-2505-43f4-aba1-0de7997b42fb)

Pair with ledger
![Screenshot 2023-11-08 at 12 22 36](https://github.com/MetaMask/metamask-mobile/assets/7315988/c3365778-d9bf-4aca-9b99-3b100fc154a7)


Transfer
![Screenshot 2023-11-08 at 13 14 58](https://github.com/MetaMask/metamask-mobile/assets/7315988/fac0dca2-7311-41c7-aec5-a1193c1962cf)

Sign message:
![Screenshot 2023-11-08 at 13 16 26](https://github.com/MetaMask/metamask-mobile/assets/7315988/23e3d58b-bae3-4fca-bd76-a80eec720653)


### **Before**



### **After**


## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

stanleyyconsensys committed:
feat: add account type tag label

stanleyyconsensys committed:
fix: revert update

stanleyyconsensys committed:
fix: update account overview

stanleyyconsensys committed:
fix: style

stanleyyconsensys committed:
fix: style

stanleyyconsensys committed:
feat: add error message to retry modal

stanleyyconsensys committed:
feat: add bluetooth library support

stanleyyconsensys committed:
fix: lint style

stanleyyconsensys committed:
fix: unit test

stanleyyconsensys committed:
remove not using package

stanleyyconsensys committed:
fix: add unit test for address list and account info

stanleyyconsensys committed:
fix: unit test on component

stanleyyconsensys committed:
fix: lint style

dawnseeker8 committed:
Merge branch 'main' into feat/add-account-type-tag

dawnseeker8 committed:
feat: fix all unit tests with more valid address.

dawnseeker8 committed:
feat: make test more valid with real eth address.

dawnseeker8 committed:
feat: unit tests, remove address util mock

dawnseeker8 committed:
Merge branch 'main' into feat/add-bluetooth-permission

dawnseeker8 committed:
Merge branch 'main' into feat/add-bluetooth-permission

dawnseeker8 committed:
Merge branch 'main' into feat/add-error-msg-to-retry-modal

dawnseeker8 committed:
Merge branch 'main' into feat/add-account-type-tag

dawnseeker8 committed:
fix: action view btn not able translate

context will not able to pass in to defaultProps, hence default translation should apply into the component itself

dawnseeker8 committed:
Merge remote-tracking branch 'origin/feat/add-account-type-tag' into feat/ledger-enable

dawnseeker8 committed:
Merge remote-tracking branch 'origin/feat/add-error-msg-to-retry-modal' into feat/ledger-enable

dawnseeker8 committed:
Merge remote-tracking branch 'origin/feat/add-bluetooth-permission' into feat/ledger-enable

dawnseeker8 committed:
feat: Working version with ledger enable
This commit didnt contain the unit tests code.

dawnseeker8 committed:
feat: Fix some missing changed files

dawnseeker8 committed:
feat: fix all unit tests and lint errors

dawnseeker8 committed:
feat: add `account_type` in some metric metric,
 also remove duplicate unit tests.

dawnseeker8 committed:
feat: add missing constant

dawnseeker8 committed:
feat: move to use ExtendedKeyringTypes

dawnseeker8 committed:
feat: use varient attribute to use theme variable

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: fix some conflict files.

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: commit to missing format of file

dawnseeker8 committed:
feat: ios Podfile.lock commit

dawnseeker8 committed:
feat: Make all account label style consistance

1. replace redentant Text to use new Terxt component
2. remove hardcoded fontSize, use theme global toker instead
3. Fix unit tests snapshot failed.

dawnseeker8 committed:
feat: Fix the unit test failure

dawnseeker8 committed:
feat: fix lint error

dawnseeker8 committed:
feat: refactory to reduce the change log in git.

dawnseeker8 committed:
feat: fix lint issue and revert change
1. Fix lint issue in RootRPCMethodsUI.js
2. Revert code in `Transactions/index.js`

dawnseeker8 committed:
feat: Revert two android setting to main version

stanleyyconsensys committed:
fix: reorder import sequence

stanleyyconsensys committed:
fix: remove un necessary change

dawnseeker8 committed:
feat: revert .project to main version.

stanleyyconsensys committed:
update address utils method sequence

stanleyyconsensys committed:
Merge branch 'feat/ledger-enable' of https://github.com/MetaMask/metamask-mobile into feat/ledger-enable

dawnseeker8 committed:
feat: Fix dedupe pipeline failure

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-enable

dawnseeker8 committed:
feat: fix ltsc issue due to enum use another value

dawnseeker8 committed:
Revert "feat: fix ltsc issue due to enum use another value"

This reverts commit 10a095fd29558c7dfe09a89a8c147e661398feaf.

dawnseeker8 committed:
feat: fix lint:tsc issue

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-enable
Also fix the issue of broken build from last commit.

dawnseeker8 committed:
feat: Fix dedupe error and one unit test.

dawnseeker8 committed:
feat: Fix dedupe ci broken.

brianacnguyen commented on app/component-library/components/Cells/Cell/foundation/CellBase/CellBase.styles.ts:
```diff
@@ -45,6 +45,7 @@ const styleSheet = (params: { theme: Theme; vars: CellBaseStyleSheetVars }) => {
     },
     tagLabel: {
       marginTop: 4,
+      height: 25,
```
please dont change this directly. instead pass the style down to the cell component

dawnseeker8 replied:
Has removed.

brianacnguyen commented on app/component-library/components/Cells/Cell/foundation/CellBase/__snapshots__/CellBase.test.tsx.snap:
```diff
@@ -206,10 +206,10 @@ exports[`CellBase should render default settings correctly 1`] = `
+            "fontWeight": "700",
             "letterSpacing": 0,
-            "lineHeight": 22,
+            "lineHeight": 20,
```
remove

dawnseeker8 replied:
after remove `LineHeight` in CellBase.Style.ts,  have regenerated the snapshot file, this file has been outdated.

brianacnguyen commented on app/component-library/components/Cells/Cell/variants/CellDisplay/__snapshots__/CellDisplay.test.tsx.snap:
```diff
@@ -229,10 +229,10 @@ exports[`CellDisplay should render default settings correctly 1`] = `
               "bottom": 2,
               "color": "#24272A",
```
remove changes

dawnseeker8 replied:
after remove `LineHeight` in CellBase.Style.ts,  have regenerated the snapshot file, this file has been outdated.

brianacnguyen commented on app/component-library/components/Cells/Cell/variants/CellMultiSelect/__snapshots__/CellMultiSelect.test.tsx.snap:
```diff
@@ -267,10 +267,10 @@ exports[`CellMultiSelect should render default settings correctly 1`] = `
+                "fontWeight": "700",
                 "letterSpacing": 0,
-                "lineHeight": 22,
+                "lineHeight": 20,
```
remove changes

brianacnguyen commented on app/component-library/components/Cells/Cell/variants/CellSelect/__snapshots__/CellSelect.test.tsx.snap:
```diff
@@ -230,10 +230,10 @@ exports[`CellSelect should render default settings correctly 1`] = `
+                "fontWeight": "700",
                 "letterSpacing": 0,
-                "lineHeight": 22,
+                "lineHeight": 20,
```
remove changes

dawnseeker8 replied:
after remove `LineHeight` in CellBase.Style.ts,  have regenerated the snapshot file, this file has been outdated.

brianacnguyen commented on app/component-library/components/Pickers/PickerAccount/PickerAccount.tsx:
```diff
@@ -58,7 +58,7 @@ const PickerAccount: React.ForwardRefRenderFunction<
         {accountTypeLabel && (
           <Text
-            variant={TextVariant.BodySM}
+            variant={TextVariant.BodySMBold}
```
did this change get approved by @Akatori-Design ?

dawnseeker8 replied:
will revert the change.

brianacnguyen commented on app/component-library/components/Tags/Tag/Tag.tsx:
```diff
@@ -17,7 +17,7 @@ const Tag = ({ label, style, ...props }: TagProps) => {
   return (
     <View style={styles.base} {...props}>
-      <Text style={styles.label} variant={TextVariant.BodyMD}>
+      <Text style={styles.label} variant={TextVariant.BodySMBold}>
```
did this change get approved by @Akatori-Design ?

dawnseeker8 replied:
the figma design for ledger-feature show the Label of devices should be `bold, font-size:12`. however because the people who create that figma screentshot didnt follow our `design-token`, i will revert this change to use standard one for label.

dawnseeker8 committed:
feat: upgrade react-native-ble-plx library

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: change code based on Brian's comments,
upgarde Podfile.lock for new react-native-ble-plx library update.

dawnseeker8 committed:
feat:update podfile.lock with react-native-ble-plx

dawnseeker8 committed:
feat: Fix `ledger` tag label, `g` has been cut off

dawnseeker8 committed:
feat: make account tag label bold

NicolasMassart commented on package.json:
```diff
@@ -342,6 +346,7 @@
     "redux-thunk": "^2.4.2",
     "reselect": "^4.0.0",
     "rn-fetch-blob": "^0.12.0",
+    "rxjs": "^7.8.1",
```
Q: what is RXJS used for in this new use case?

dawnseeker8 replied:
`useBluetooth.ts` and `useBluetoothDevices.ts` has use `rxjs`

NicolasMassart reviewed (commented):
I've reviewed :
- constants
- core/Analytics
- core/Engine.ts
- images
- util/test/testSetup.js
- ios
- package.json

gantunesr commented on patches/@metamask+keyring-controller+6.1.0.patch:
```diff
@@ -1,18 +1,73 @@
+diff --git a/node_modules/@metamask/keyring-controller/dist/KeyringController.d.ts b/node_modules/@metamask/keyring-controller/dist/KeyringController.d.ts
+index dff777b..bb3aad2 100644
```
@stanleyyconsensys @dawnseeker8 what are the specific changes to the keyring-controller? This patch seems odd

dawnseeker8 replied:
Those strange character look like caused by sourcemap, we will find a way to regenerate the patch without .map files to simplify the patch contents.

dawnseeker8 replied:
hi @gantunesr i have checked some of our patches files, such as `@metamask+approval-controller+3.4.0.patch` and `@metamask+signature-controller+4.0.1.patch`, they also have those strange characters for their `.map` source mapping file change. should we still want to tackle `@metamask+keyring-controller+6.1.0.patch` by removing those source map diff?

dawnseeker8 replied:
hi, @gantunesr i have manually remove those source file diff in our patch, and also test it was working ok, we can resolve this now. :)

brianacnguyen commented on app/component-library/components/Tags/Tag/Tag.styles.ts:
```diff
@@ -25,7 +25,7 @@ const styleSheet = (params: { theme: Theme; vars: TagStyleSheetVars }) => {
         borderWidth: 1,
         borderRadius: 10,
-        height: 20,
+        height: 22,
```
why was this changed?

dawnseeker8 replied:
After discuss with Brian, we design to revert the change, and log a defect for `ledger` tag label `g` being cut off for brian to look at.

dawnseeker8 committed:
feat: revert the tag style to `main` version.

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable
Fix the screen UI issue in send token screen.

dawnseeker8 committed:
feat: fix lint issue from rebase.

dawnseeker8 committed:
feat: remove unused images.

dawnseeker8 commented on app/images/etherscan-dark.png:
I double check the source code, and didn't find any where use this image, i have removed this image in my last commit.

dawnseeker8 commented on app/images/etherscan-light.png:
I double check the source code, and didn't find any where use this image, i have removed this image in my last commit.

vivek-consensys commented:
Manual Ledger Test Flows:- 

1. Connect to Ledger, import tokens on the Ledger account

https://github.com/MetaMask/metamask-mobile/assets/106310394/13419402-56bb-4699-b0df-4b3e7164b580


2. Disconnect to Ledger


https://github.com/MetaMask/metamask-mobile/assets/106310394/19f6f4b6-8bb3-4f2f-b0d3-dafe1c14ac1a

dawnseeker8 committed:
feat: fix the ios build issue in britise

dawnseeker8 committed:
feat: Manually remove the source map diff in patch
currently it is working version after testing.

vivek-consensys commented:
Manual Ledger Test Flows:-
3. Send transaction from Ledger 

https://github.com/MetaMask/metamask-mobile/assets/106310394/b6d18f66-6e9b-47cf-9c52-357689a48276

vivek-consensys commented:
Manual Ledger Test Flows:-
4. Create & Approve Token using MetaMask Test-dApp 

https://github.com/MetaMask/metamask-mobile/assets/106310394/7d11e581-de31-4c57-8de3-7593bdb99f5d

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: Revert javascriptEnabled to true in webview
1. 31e5f39 commit cause test-dapp misbehave,

dawnseeker8 committed:
feat: revert cocapods version to `main` version.

dawnseeker8 committed:
feat: fix lint issue from rebase code.

dawnseeker8 committed:
feat:add comment description into patch file
add some more information in patch to discribe the purpose of this patch

dawnseeker8 committed:
feat: Fix the syntax error in Podfile.lock

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: commit the missing file from last commit

matthewwalsh0 commented on app/core/RPCMethods/RPCMethodMiddleware.ts:
```diff
@@ -888,6 +890,19 @@ export const getRpcMethodMiddleware = ({
     }
-    await rpcMethods[req.method]();
+
+    const isWhiteListedMethod = isWhitelistedRPC(req.method);
```
As this is stable and generic, this is great if it facilitates what is needed for this first implementation.

But I'd also be very happy to discuss how we can improve the confirmation architecture or signature controller if there is an aspect that is missing to support the Ledger use case.

Forgive me if this isn't relevant but the approval requests as they are know should throw in the backend if the approval is rejected, but equally we have a mechanism called result callbacks that can reject a promise held in the the frontend to notify it of any backend failures.

dawnseeker8 replied:
hi, @matthewwalsh0 thank you for suggestion, we will be happy to discuss with you after this PR merged to see any improvement we can do in core SignatureController to improve this step. :)

matthewwalsh0 commented:
No issues from the confirmations system perspective, except for potential future improvements to the error flows in case these can be "natively" supported by the `ApprovalController` and `SignatureController`.

Have also forwarded the PR to @jpuri to keep in the loop concerning the frontend confirmation changes.

gantunesr commented on app/components/UI/LedgerModals/Steps/SearchingForDeviceStep.tsx:
```diff
@@ -0,0 +1,121 @@
+    navigation.push('Webview', {
+      screen: 'SimpleWebview',
+      params: {
+        url: 'https://support.ledger.com/hc/en-us/articles/360009576554-Ethereum-ETH-?docs=true',
```
Move this url to a constants file (`app/constants/urls.ts`).

dawnseeker8 replied:
have moved 'https://support.ledger.com/hc/en-us/articles/360009576554-Ethereum-ETH-?docs=true` into urls.ts in constant folder.

gantunesr commented on app/components/UI/Swaps/QuotesView.js:
```diff
@@ -882,7 +886,7 @@ function SwapsQuotesView({
           network_fees_ETH: renderFromWei(toWei(selectedQuoteValue?.ethFee)),
           chain_id: chainId,
         };
-        Analytics.trackEventWithParameters(MetaMetricsEvents.SWAP_STARTED, {});
+
         Analytics.trackEventWithParameters(
```
Is there a reason why you removed this event?

dawnseeker8 replied:
hi, @gantunesr if you look at the code here, we log a `SWAP_STARTED` with empty information first , and then immediately we log a same `SWAP_STARTED` with swap information. if we dont remove that empty line, metric will log twice with one empty information and one not empty information. if you think that line is still needed, i can revert the `removed` code in line 885.

dawnseeker8 replied:
i have reverted that line 885 in the latest commit.

gantunesr commented on app/components/Views/LedgerConnect/index.tsx:
```diff
@@ -0,0 +1,337 @@
+    navigation.push('Webview', {
+      screen: 'SimpleWebview',
+      params: {
+        url: 'https://support.ledger.com/hc/en-us/articles/360009576554-Ethereum-ETH-?docs=true',
```
Use the constant value here

dawnseeker8 replied:
has changed to use Urls. constant.

gantunesr commented on app/components/Views/LedgerConnect/index.tsx:
```diff
@@ -0,0 +1,337 @@
+      {errorDetail && <LedgerConnectionError {...errorDetail} />}
+    </SafeAreaView>
+  );
+};
+// Emanuel Charlie Tiparu
+export default React.memo(LedgerConnect);
```
What's this comment? Lets remove it

```suggestion
};

export default React.memo(LedgerConnect);
```

gantunesr commented on app/components/Views/SendFlow/Confirm/index.js:
```diff
@@ -661,6 +665,49 @@ class Confirm extends PureComponent {
+    gaParams,
+  ) => {
+    const { TransactionController } = Engine.context;
+    const { navigation } = this.props;
+    //manual cancel from UI / rejected from ledger
+    try {
+      if (!approve) {
```
nit

```suggestion
    const { navigation } = this.props;
    // Manual cancel from UI or rejected from ledger device.
    try {
      if (!approve) {
```

gantunesr commented on app/components/Views/SendFlow/Confirm/index.js:
```diff
@@ -661,6 +665,49 @@ class Confirm extends PureComponent {
+        });
+      }
+    } finally {
+      //Let ledger modal handle the error
```
Nit
```suggestion
      // Error handling derived to LedgerConfirmationModal component
```

gantunesr commented on app/components/hooks/Ledger/useBluetooth.test.ts:
```diff
@@ -0,0 +1,11 @@
+import { renderHook } from '@testing-library/react-hooks';
+import useBluetooth from './useBluetooth';
+
+describe('useBluetooth', () => {
```
In a next PR we should improve the testing around the bluetooth hooks

dawnseeker8 replied:
we will improve this test in next PR, currently for testing this, most difficult part is mock the `@ledgerhq/react-native-hw-transport-ble`, and following code in `useBluetooth.ts`

```
 import('@ledgerhq/react-native-hw-transport-ble').then(
        (BluetoothTransport: any) => {
          subscription = BluetoothTransport.default.observeState({
            next: (e: { available: boolean; type: State }) => {
              if (e.available && e.type === State.PoweredOn && !bluetoothOn) {
                setBluetoothOn(true);
                setBluetoothConnectionError(false);
              }

              if (!e.available && e.type === State.PoweredOff) {
                setBluetoothOn(false);
                setBluetoothConnectionError(true);
              }
            },
          });
        },
      );
      ```

gantunesr commented on app/constants/navigation/Routes.ts:
```diff
@@ -18,6 +18,16 @@ const Routes = {
+  HW: {
+    CONNECT: 'ConnectHardwareWalletFlow',
+    SELECT_DEVICE: 'SelectHardwareWallet',
+    CONNECT_QR_DEVICE: 'ConnectQRHardwareFlow',
+    CONNECT_LEDGER: 'LedgerConnectFlow',
```
Nit
```suggestion
    CONNECT_QR_DEVICE: 'ConnectQRHardwareFlow',
    CONNECT_LEDGER: 'ConnectLedgerFlow',
```

gantunesr commented on app/core/Ledger/Ledger.ts:
```diff
@@ -0,0 +1,202 @@
+} from '@consensys/ledgerhq-metamask-keyring';
+import ExtendedKeyringTypes from '../../constants/keyringTypes';
+
+//TODO - Patch EthKeyringController into KeyringController
```
Is this the same patch in this PR or another one?

dawnseeker8 replied:
Stanely has remove the `TODO` comment, current patch in PR is sufficient enought.

stanleyyconsensys committed:
fix: update ledger.ts

- not exporting getEthKeyringController
- consistently getting ethKeyringController from local method  getEthKeyringController

dawnseeker8 committed:
Apply suggestions from code review

Apply Gustova suggestion for code change.

Co-authored-by: Gustavo Antunes <17601467+gantunesr@users.noreply.github.com>

dawnseeker8 commented on app/components/UI/LedgerModals/LedgerConfirmationModal.tsx:
```diff
@@ -0,0 +1,297 @@
+/* eslint-disable @typescript-eslint/no-require-imports */
+/* eslint-disable import/no-commonjs */
+/* eslint-disable @typescript-eslint/no-var-requires */
```
have been removed!

dawnseeker8 committed:
feat: refactory code based on Custova's comment

dawnseeker8 commented on app/components/UI/LedgerModals/LedgerConfirmationModal.tsx:
```diff
@@ -0,0 +1,297 @@
+        await unlockLedgerDefaultAccount();
+        await onConfirmation();
+      });
+    } catch (_e) {
+      // Handle a super edge case of the user starting a transaction with the device connected
+      // After arriving to confirmation the ETH app is not installed anymore this causes a crash.
+      AnalyticsV2.trackEvent(MetaMetricsEvents.LEDGER_HARDWARE_WALLET_ERROR, {
+        device_type: 'Ledger',
+        error: 'LEDGER_ETH_APP_NOT_INSTALLED',
+      });
+    }
```
this was very old code from ledger team, currently we can't make this throw statement occur, after talking to Stanley , he suggest we keep this code and aslo add a log here to track whether any users hit this throw, and try to identify how this could happen.

gantunesr replied:
Is this a product metric we're interested? If not, it's better to log it into Sentry for dev purposes

dawnseeker8 replied:
this is product metric listed in https://docs.google.com/document/d/1sBeYy-pA9cpXdD-jy48CoZTYFusSdWkm2kM5WDClybc/edit#heading=h.mq7tghwjdb93

dawnseeker8 committed:
feat: revert remove line.

dawnseeker8 committed:
feat: Revert the files was commited by mistaken.

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

NicolasMassart reviewed (commented):
It looks good to me for the files I listed as reviewed in my previous comment

plasmacorral commented:
I have some concerns here after some manual testing on iOS 17.0.3 with iPhone Xs using a Nano X with eth app 1.10.1 and blind signing enabled.

1) Signatures are not being returned to the [test dapp](https://metamask.github.io/test-dapp) on Ledger in this branch, but Keystone works just fine. 
  Keystone successfully using Eth sign and Personal sign on the test dapp: https://recordit.co/ozWHGC0DwX
  Ledger signing attempts: https://recordit.co/PA4g8SH0jI
2) After binding a Keystone address to MMM, user is returned to hardware wallet selection screen.  On successful binding with Keystone, user should be returned to Wallet view with the freshly added Keystone address selected. 
  https://recordit.co/9hnHZDV11s 
3) Was able to pair a Ledger device without clicking anything on the personal security device (PSD).  I selected to pair from MMM and it just happened in this branch.  In testing from 2022 I was required to tap on the PSD to confirm that I wished to pair to MMM.  This is not happening today and could be a security issue: https://recordit.co/ns7WBIWaVF

plasmacorral reviewed (dismissed):
https://github.com/MetaMask/metamask-mobile/pull/7757#issuecomment-1819992123

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: change based on Mike comments
after user unlock keyring accounts, screen will direct to Wallet View

dawnseeker8 commented:
hi, @plasmacorral, 
I have fixed the point 2 issue, now the keyring accounts selected, screen will be redirect to Wallet view.

dawnseeker8 committed:
feat: Revert all javascript disabled code
This commit is to check whether Ios will work without Tomas's code.

vivek-consensys commented:
> I have some concerns here after some manual testing on iOS 17.0.3 with iPhone Xs using a Nano X with eth app 1.10.1 and blind signing enabled.
> 
> Android device is Pixel 5a with Android 14.
> 
> Tested with a QA build of commit [8778e77](https://github.com/MetaMask/metamask-mobile/commit/8778e7789dc12d7b25985a70e02114f85c724632)
> 
> 1. Signatures are not being returned to the [test dapp](https://metamask.github.io/test-dapp) on Ledger in this branch, but Keystone works just fine. I had no issues conducting a send tx with the test dapp on iOS.  Did not have this issue on Android.  Can confirm this impedes the iOS user's ability to sign a message to login to Opensea, for example.
>    Keystone successfully using Eth sign and Personal sign on the test dapp: https://recordit.co/ozWHGC0DwX
>    Ledger signing attempts on test dapp (eth sign/personal sign): https://recordit.co/PA4g8SH0jI
>    Three attempts to login to Opensea with Ledger + MMM: https://recordit.co/CVmITcwCEq
> 2. After binding a Keystone address to MMM, user is returned to hardware wallet selection screen on both android and iOS.  On successful binding with Keystone, user should be returned to Wallet view with the freshly added Keystone address selected.
>    https://recordit.co/9hnHZDV11s
> 3. ~Was able to pair a Ledger device without clicking anything on the personal security device (PSD).  I selected to pair from MMM and it just happened in this branch without any consent on the Ledger.  In testing from 2022 I was required to tap on the PSD to confirm that I wished to pair to MMM.  Did not have this issue on Android.  This is not happening today with my iOS device and could be a security issue: https://recordit.co/ns7WBIWaVF~. This was only the case because Ledger had been previously paired with iPhone.  Once I forgot the prior bt connection, I was prompted to pair on the Ledger PSD as expected. This was testing error and is not an issue.

Test Flow Screen Recordings:- 

Manual testing using iPhone 13 on iOS 17.1
Tested with a QA build of commit https://app.bitrise.io/build/0dc22016-e6b8-4858-b88f-b11aa53a84c3?tab=artifacts

1. Signing using MetaMask test-dApp

https://github.com/MetaMask/metamask-mobile/assets/106310394/3da9b751-28d3-4f99-93f5-2107e5d0cc7c

2. Signing using OpenSea

https://github.com/MetaMask/metamask-mobile/assets/106310394/8e652a0e-8cc2-4840-bc4e-12b784b51690

3. User returned to wallet view after connecting to Keystone wallet

https://github.com/MetaMask/metamask-mobile/assets/106310394/841e4239-a11a-44ac-a65e-1e6e160677c2

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

plasmacorral commented:
Issue #1 resolved according to videos above, Thank you @vivek-consensys!

I have one last nit that is a correction to an earlier error I made.  For #2 above I made a mistake and missed the fact that in production today after adding or removing hardware, the user is returned to wallet view **_with the accounts drop-down menu presented._**  

As tested today with Pixel 5a on Android 14 and iPhone Xs with iOS 17.0.3 with a QA build of commit bbdee57ae94c0721f5a2321514a2b6524026a4f9.

Here is a grid that includes recordings comparing production results to the branch under test: 
<google-sheets-html-origin><style type="text/css"><!--td {border: 1px solid #cccccc;}br {mso-data-placement:same-cell;}--></style>
  | bbdee57 commit lands | Commit bbdee57 video | Production lands | Production v7.11.0 build 1197
-- | -- | -- | -- | --
Adding Ledger | Wallet View | https://recordit.co/JHt9OwQn1J |   |  
Forgetting Ledger | Hardware selection.  When I tap back arrow I returned to Wallet view with accounts drop-down open | https://recordit.co/7YC1AkeFEG |   |  
Adding Keystone | Wallet View | https://recordit.co/XqIm4TW7d0 | Wallet view with accounts drop-down open | https://recordit.co/eL0dDcYz9e
Forgetting Keystone | Hardware selection.  When I tap back arrow I returned to Wallet view with accounts drop-down open | https://recordit.co/3WKzk0tIR4 | Wallet view with accounts drop-down open | https://recordit.co/odGKM2vowS

plasmacorral commented:
I can otherwise report that signing and sending with the test dapp from both Keystone and Ledger (when both present in the wallet) looks good in commit bbdee57ae94c0721f5a2321514a2b6524026a4f9 tested with Pixel 5a on Android 14 and iPhone Xs with iOS 17.0.3 using a QA build

gantunesr commented:
[Running E2E](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/871803b2-40c4-4677-b448-981753248c0c)

gantunesr commented:
@SocketSecurity ignore-all

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: change QR connect to account drawdown list screen

dawnseeker8 commented:
> Issue #1 resolved according to videos above, Thank you @vivek-consensys!
> 
> I have one last nit that is a correction to an earlier error I made. For two above I made a mistake and missed the fact that in production today after adding or removing hardware, the user is returned to wallet view **_with the accounts drop-down menu presented._**
> 
> As tested today with Pixel 5a on Android 14 and iPhone Xs with iOS 17.0.3 with a QA build of commit [bbdee57](https://github.com/MetaMask/metamask-mobile/commit/bbdee57ae94c0721f5a2321514a2b6524026a4f9).
> 
> Here is a grid that includes recordings comparing production results to the branch under test: <style type="text/css"></style>
> 
>  	[bbdee57](https://github.com/MetaMask/metamask-mobile/commit/bbdee57ae94c0721f5a2321514a2b6524026a4f9) commit lands	Commit [bbdee57](https://github.com/MetaMask/metamask-mobile/commit/bbdee57ae94c0721f5a2321514a2b6524026a4f9) video	Production lands	Production v7.11.0 build 1197
> Adding Ledger	Wallet View	https://recordit.co/JHt9OwQn1J	 	 
> Forgetting Ledger	Hardware selection. When I tap back arrow I returned to Wallet view with accounts drop-down open	https://recordit.co/7YC1AkeFEG	 	 
> Adding Keystone	Wallet View	https://recordit.co/XqIm4TW7d0	Wallet view with accounts drop-down open	https://recordit.co/eL0dDcYz9e
> Forgetting Keystone	Hardware selection. When I tap back arrow I returned to Wallet view with accounts drop-down open	https://recordit.co/3WKzk0tIR4	Wallet view with accounts drop-down open	https://recordit.co/odGKM2vowS

I have changed the code to make `Adding Keystone Wallet View` and `Forgetting Keystone Hardware Selection` behave same as `Production v7.11.0 build 1197` (same as `main` as well), now the screen will direct to `Wallet view with accounts drop-down open`.  for `adding ledger`, i also change the code to direct to `Wallet view with accounts drop-down open` to make behave consistance.

dawnseeker8 committed:
Merge branch 'main' into feat/ledger-enable

dawnseeker8 committed:
feat: remove duplicate permission.

dawnseeker8 committed:
feat: fix the typo error in text attribute.

dawnseeker8 committed:
feat: Try to fix the ios e2e tests

dawnseeker8 committed:
feat: After few testing, confirm AccountView is fine.

dawnseeker8 commented:
attached new e2e tests result [here](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/ce00e5a6-2d9f-44a7-b6d7-b4623a694366)

stanleyyconsensys commented on app/constants/keyringTypes.ts:
```diff
@@ -1,3 +1,8 @@
+  qr = 'QR Hardware Wallet Device',
+  ledger = 'Ledger Hardware',
+}
+
```
app/util/importAdditionalAccounts.js using own keyring type text, it should replace by ExtendedKeyringTypes.hd

dawnseeker8 replied:
has changed  `importAdditionalAccounts ` based o your comments.

dawnseeker8 committed:
feat: refactory to remove unuse outdating code.

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: add retry to prevent user immediately see error screen with eth app not open.

dawnseeker8 committed:
feat: Commit based on Stanley feedback.

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: Dont store rpcEvents and accounts state in localStorage.

dawnseeker8 committed:
feat: try to fix regression e2e tests signing model not popup.

dawnseeker8 committed:
feat: check whether fix ios gem install error in ios e2e tests.

dawnseeker8 committed:
feat: Check whether e2e ios build can pass in bitrise or not.

dawnseeker8 committed:
feat: Revert previous two ios change to `main` version.

dawnseeker8 committed:
feat: Revert the suspect line which may break ios e2e build.

dawnseeker8 committed:
feat: revert the change to match main

gantunesr committed:
fea: hide ledger (#7940)

This change hides the ledger feature behind an experimental flag.

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: to test e2e test pass or not.

dawnseeker8 committed:
feat: Implement option 1 to turn off ledger feature..

dawnseeker8 committed:
Merge remote-tracking branch 'origin/main' into feat/ledger-enable

dawnseeker8 committed:
feat: revert the code to disable the ledger feature.

dawnseeker8 committed:
feat: revert file to main version

dawnseeker8 committed:
feat: revert again to remove the change from main

dawnseeker8 committed:
feat: Remove the duplicate lines in `Podfile.lock` after rebase which breank ios build.

dawnseeker8 commented:
This is the latest smoke e2e tests result https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/74ad83d1-aaea-4065-935b-2d6c220e5d0d

dawnseeker8 commented:
This is the latest regression e2e tests result: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/4553befb-80f7-4c38-883b-7e9a95d6943a

gantunesr commented on .detoxrc.js:
```diff
@@ -23,7 +23,7 @@ module.exports = {
       },
     },
-  },
+  }, 
```
Nit: Looks like a space got added in this line

dawnseeker8 replied:
Already resolved in my last commit to revert the version to match main version.

gantunesr commented on .detoxrc.js:
```diff
@@ -10,7 +10,7 @@ module.exports = {
     retries: 2,
   },
-
+  
```
Nit: Same here

dawnseeker8 replied:
Already resolved in my last commit to revert the version to match main version.

gantunesr reviewed (approved):
I've manually tested the PR and verified the test cases

Akaryatrh commented:
I've manually tested the PR and verified the test cases

georgewrmarshall reviewed (approved):
Approving as @brianacnguyen previously mentioned no updates to component-library folder
